manticore/platforms/evm_world_state.py

```diff
@@ -0,0 +1,478 @@
+    result of the call to get_storage_data was cached in the overlay, then no exception would be
+    thrown.
+    """
+
+    def __init__(self, underlay: WorldState):
+        self._underlay: WorldState = underlay
+        self._deleted_accounts: Set[int] = set()
+        self._nonce: Dict[int, Union[int, BitVec]] = {}
+        self._balance: Dict[int, Union[int, BitVec]] = {}
+        self._storage: Dict[int, Storage] = {}
+        self._code: Dict[int, Union[bytes, Array]] = {}
+        self._blocknumber: Optional[Union[int, BitVec]] = None
+        self._timestamp: Optional[Union[int, BitVec]] = None
+        self._difficulty: Optional[Union[int, BitVec]] = None
+        self._gaslimit: Optional[Union[int, BitVec]] = None
+        self._coinbase: Optional[Union[int, BitVec]] = None
+
+    def is_remote(self) -> bool:
+        return self._underlay.is_remote()
+
+    def accounts(self) -> Set[int]:
+        accounts: Set[int] = set()
+        try:
+            accounts = self._underlay.accounts()
+        except NotImplementedError:
+            pass
+        return (
+            accounts
+            | self._nonce.keys()
+            | self._balance.keys()
+            | self._storage.keys()
+            | self._code.keys()
+        )
+
+    def get_nonce(self, address: int) -> Union[int, BitVec]:
+        if address in self._nonce:
+            return self._nonce[address]
+        else:
+            return self._underlay.get_nonce(address)
+
+    def get_balance(self, address: int) -> Union[int, BitVec]:
+        if address in self._balance:
+            return self._balance[address]
+        else:
+            return self._underlay.get_balance(address)
+
+    def has_storage(self, address: int) -> bool:
+        dirty = False
+        try:
+            dirty = self._underlay.has_storage(address)
+        except NotImplementedError:
+            pass
+        storage = self._storage.get(address)
+        if storage is not None:
+            dirty = dirty or storage.dirty
+        return dirty
+
+    def get_storage(self, address: int) -> Optional[Storage]:
+        storage = None
+        try:
+            storage = self._underlay.get_storage(address)
+        except NotImplementedError:
+            pass
+        # sam.moelius: Rightfully, the overlay's storage should be merged into the underlay's
+        # storage.  However, this is not currently implemented.
+        if storage is not None:
+            # sam.moelius: At present, the get_storage methods of both DefaultWorldState and
+            # RemoteWorldState raise NotImplementedError.  So this exception should be unreachable.
+            raise NotImplementedError("Merging of storage is not implemented")
+        storage = self._storage.get(address)
+        return storage
+
+    def get_storage_data(
+        self, constraints: ConstraintSet, address: int, offset: Union[int, BitVec]
```

Comment:
note to self. Why do we need the constraints here

Reply:
Technically, no, but: https://github.com/trailofbits/manticore/blob/eecd9b5a67b25e2cabe55ffc783e7a0ea65d7bde/manticore/platforms/evm_world_state.py#L106-L107

Reply:
At one point I was toying with the idea of creating a new `Storage` here:
https://github.com/trailofbits/manticore/blob/eecd9b5a67b25e2cabe55ffc783e7a0ea65d7bde/manticore/platforms/evm_world_state.py#L383-L384

Reply:
Why get_storage_data is responsible of creating new "Storage" ?

We learn that the account need an Storage when it is actually read?

Reply:
My concern is: will there be cases where we want to make assertions about a contract's Storage before it has been written to?  By "make assertions," I mean something like this:
https://github.com/trailofbits/manticore/blob/55e2486f4791757b77fd16b0dfe65d4ecb6e3e80/tests/other/test_smtlibv2.py#L146-L148
Or, more concretely, would we want to allow something like this?
```
    cs.add(world_state.get_storage_data(address, offset) == ord('A'))
```
Right now, this won't work if `address`'s Storage has not been written to.

I *think* we want this to not work. We currently take a "closed world" approach in the sense that the only addresses we acknowledge in the overlay are those to which we have observed writes. While this may sound restrictive, it actually facilitates solving certain types of constraint sets. An example is here:
https://github.com/trailofbits/manticore/blob/eecd9b5a67b25e2cabe55ffc783e7a0ea65d7bde/tests/ethereum/test_storage.py#L235-L238
That `assertFalse` would fail if we did not have this "closed world" assumption.

Having said that, there is a tiny voice in my head that says this may not be the right approach, i.e., there may be situations where we want to allow assertions on a contract's Storage, even though we have not observed writes to it.

Adding to this, `set_storage_data` requires a `constraints` parameter. So requiring it also for `get_storage_data` does not seem that jarring.

Reply:
3 subcases here:
    ** We know all the already used in the node
    ** We know some of the used slots in the node
    ** We do not know any of the already used slots in the node


Lets assume it is not possible to get all known slots from a storage in the node. :( Booo
So the known slots could start to be just all non zero storage items from offsets 0 to 100 ?

```known_slots = (0,1,2,3)```

This will account for the static variables. Maybe that's enough?
Other static analysis over the bytecode can be done. We get all those and write them to our model.


So initially when an unknown account is referenced (as a tx target) we bring it 
from the node and build an OverlayedRemoteWorld and storage. We create the overlay 
account with an overlayed array storage and write the known values to it (The ones from 0 to 100 for now?).

```
class OStorage:
    array :ArrayProxy
    node :Node
```

READ
====
Now suppose that storage is then read at a free symbolic index: index

oStorage.read(index) -> ?????

I can imagine 2 things to happen here. And it may even request a fork if both are possible at the same time.

* index is known.

The ArrayProxy put in front of the node knows about the index. `array.is_known(index) -> True`.
The result of reading here is simply `array[index]`. We do not need to use a default as we know the index is known.
This returns a symbol representing one value read from the full _known_ storage.


* index is not known 

The ArrayProxy put in front of the node does not knows about the index. `array.is_known(index) -> False`.
We need to look for the value in the node.
But where? The index is symbolic. Which actual storage slot do we bring down from the node?
This sucks.
Suck 1: Ask the solver and sample solutions for index. (note there are no _known_ yet)
        Get the corresponding values from the node. 
        Print a warning about incompletness (if not ALL solutions where retrieved)
        Write them to the array
        Constrain index to be one of the solutions
        Index is now _known_ then go to the previous case

Suck 2:
        Assume all storage in the node above 100 are blank
        Go to previous case

WRITE
=====
Writing to a storage slot requires reading it first. So it will resolve itself when solving reading